tests/product/image_builder.py

```diff
@@ -0,0 +1,203 @@
+from subprocess import check_output
+from sys import argv, exit
+from tests.configurable_cluster import ConfigurableCluster
+from tests.docker_cluster import DockerCluster, DockerClusterException
+from tests.no_hadoop_bare_image_provider import NoHadoopBareImageProvider
+
+from tests.product.mode_installers import StandaloneModeInstaller, \
+    YarnSliderModeInstaller
+from tests.product.prestoadmin_installer import PrestoadminInstaller
+from tests.product.standalone.presto_installer import StandalonePrestoInstaller
+from tests.product.topology_installer import TopologyInstaller
+from tests.product.yarn_slider.slider_installer import SliderInstaller
+
+import argparse
+
+
+class ProductTestImageBuilder:
+    STANDALONE_BARE_CLUSTER = 'bare'
+    BARE_CLUSTER = 'bare'
+    PA_ONLY_CLUSTER = 'pa_only_standalone'
+    STANDALONE_PRESTO_CLUSTER = 'presto'
+
+    PA_ONLY_YS_CLUSTER = 'pa_only_ys'
+    PA_SLIDER_CLUSTER = 'pa_slider'
+
+    _cluster_types = {
+        BARE_CLUSTER: [],
+        PA_ONLY_CLUSTER: [PrestoadminInstaller,
+                          StandaloneModeInstaller],
+        STANDALONE_PRESTO_CLUSTER: [PrestoadminInstaller,
+                                    StandaloneModeInstaller,
+                                    TopologyInstaller,
+                                    StandalonePrestoInstaller],
+        PA_ONLY_YS_CLUSTER: [PrestoadminInstaller,
+                             YarnSliderModeInstaller],
+        PA_SLIDER_CLUSTER: [PrestoadminInstaller,
+                            YarnSliderModeInstaller,
+                            SliderInstaller]
+    }
+
+    def __init__(self):
+        self.maxDiff = None
+        self.cluster = None
+        self.default_keywords = {}
+
+    def _apply_post_install_hooks(self, installers):
+        for installer in installers:
+            self.cluster.postinstall(installer)
+
+    def _update_replacement_keywords(self, installers):
+        for installer in installers:
+            installer_instance = installer(self)
+            self.default_keywords.update(installer_instance.get_keywords())
+
+    def _run_installers(self, installers):
+        cluster = self.cluster
+        for installer in installers:
+            dependencies = installer.get_dependencies()
+
+            for dependency in dependencies:
+                dependency.assert_installed(self)
+
+            installer_instance = installer(self)
+            installer_instance.install()
+
+            self.default_keywords.update(installer_instance.get_keywords())
+            cluster.postinstall(installer)
+
+    def _setup_cluster(self, bare_image_provider, cluster_type):
+        installers = None
```

Comment:
It had to be there because of the try-except block. I can get rid of because of the argparsing:
>I would remove this try except for KeyError because this validation happens implicitly at the argument parsing level.